src/particles.zig

```diff
@@ -0,0 +1,413 @@
+		emissionTextureArray.generate(emissionTextures.items, true, false);
+		c.glTexParameterf(c.GL_TEXTURE_2D_ARRAY, c.GL_TEXTURE_MAX_ANISOTROPY, @floatFromInt(main.settings.anisotropicFiltering));
+
+		particleTypesSSBO = SSBO.initStatic(ParticleType, ParticleManager.types.items);
+		particleTypesSSBO.bind(14);
+	}
+};
+
+pub const ParticleSystem = struct {
+	pub const maxCapacity: u32 = 524288;
+	var particleCount: u32 = 0;
+	var particles: [maxCapacity]Particle = undefined;
+	var particlesLocal: [maxCapacity]ParticleLocal = undefined;
+	// TODO: add different emitters for different types of movements like windy, normal, no collisions and etc.
+	var properties: EmitterProperties = undefined;
+	var previousPlayerPos: Vec3d = undefined;
+
+	var particlesSSBO: SSBO = undefined;
+
+	var pipeline: graphics.Pipeline = undefined;
+	const UniformStruct = struct {
+		projectionAndViewMatrix: c_int,
+		billboardMatrix: c_int,
+		ambientLight: c_int,
+	};
+	var uniforms: UniformStruct = undefined;
+
+	pub fn init() void {
+		std.log.debug("Particle alignment: {d} size: {d}\n", .{@alignOf(Particle), @sizeOf(Particle)});
+		pipeline = graphics.Pipeline.init(
+			"assets/cubyz/shaders/particles/particles.vert",
+			"assets/cubyz/shaders/particles/particles.frag",
+			"",
+			&uniforms,
+			.{},
+			.{.depthTest = true, .depthWrite = true},
+			.{.attachments = &.{.noBlending}},
+		);
+
+		properties = EmitterProperties{
+			.gravity = .{0, 0, -2},
+			.drag = 0.5,
+			.lifeTimeMin = 5,
+			.lifeTimeMax = 5,
+			.velMin = 0.1,
+			.velMax = 0.3,
+			.rotVelMin = std.math.pi*0.2,
+			.rotVelMax = std.math.pi*0.6,
+		};
+		particlesSSBO = SSBO.init();
+		particlesSSBO.createDynamicBuffer(Particle, maxCapacity);
+		particlesSSBO.bind(13);
+
+		seed = @bitCast(@as(i64, @truncate(std.time.nanoTimestamp())));
+	}
+
+	pub fn deinit() void {
+		pipeline.deinit();
+		particlesSSBO.deinit();
+	}
+
+	pub fn update(deltaTime: f32) void {
+		const vecDeltaTime: Vec4f = @as(Vec4f, @splat(deltaTime));
+		const playerPos = game.Player.getEyePosBlocking();
+		const prevPlayerPosDifference: Vec3f = @floatCast(previousPlayerPos - playerPos);
+
+		var i: u32 = 0;
+		while(i < particleCount) {
+			const particle = &particles[i];
+			const particleLocal = &particlesLocal[i];
+			particle.lifeRatio -= particleLocal.lifeVelocity*deltaTime;
+			if(particle.lifeRatio < 0) {
+				particleCount -= 1;
+				particles[i] = particles[particleCount];
+				particlesLocal[i] = particlesLocal[particleCount];
+				continue;
+			}
+
+			var rot = particle.posAndRotation[3];
+			const rotVel = particleLocal.velAndRotationVel[3];
+			rot += rotVel*deltaTime;
+
+			particleLocal.velAndRotationVel += vec.combine(properties.gravity, 0)*vecDeltaTime;
+			particleLocal.velAndRotationVel *= @splat(@max(0, 1 - properties.drag*deltaTime));
+			const posDelta = particleLocal.velAndRotationVel*vecDeltaTime;
+
+			// TODO: OPTIMIZE THE HELL OUT OF THIS
```

Comment:
Unprofessional wording, code should not contain jokes so it is easy to read. Also this should be an issue.

Reply:
Yeah, please make an issue for it. As for the exact wording, I think a little fun in comments is nice sometimes.

Reply:
also next time don't forget to link it here: #1471